interpreter/golabels/tls_amd64.go

```diff
@@ -6,43 +6,95 @@
 package golabels // import "go.opentelemetry.io/ebpf-profiler/interpreter/golabels"
 
 import (
-	log "github.com/sirupsen/logrus"
+	"errors"
+
 	"go.opentelemetry.io/ebpf-profiler/libpf/pfelf"
+	"go.opentelemetry.io/ebpf-profiler/nativeunwind/elfunwindinfo"
 	"golang.org/x/arch/x86/x86asm"
 )
 
+var (
+	errMissingSymbol = errors.New("failed to find runtime.stackcheck")
+	errUnexpectedAsm = errors.New("failed to disassemble runtime.stackcheck")
+)
+
+// virtualMemoryReader allows to mock pfelf.File for testing.
+type virtualMemoryReader interface {
+	VirtualMemory(addr int64, size int, align int) ([]byte, error)
+}
+
 // Most normal amd64 Go binaries use -8 as offset into TLS space for
 // storing the current g but "static" binaries it ends up as -80. There
 // may be dynamic relocating going on so just read it from a known
 // symbol if possible.
 func extractTLSGOffset(f *pfelf.File) (int32, error) {
-	syms, err := f.ReadSymbols()
+	pclntab, err := elfunwindinfo.NewGopclntab(f)
 	if err != nil {
 		return 0, err
 	}
+	defer pclntab.Close()
+
 	// Dump of assembler code for function runtime.stackcheck:
-	// 0x0000000000470080 <+0>:     mov    %fs:0xfffffffffffffff8,%rax
-	sym, err := syms.LookupSymbol("runtime.stackcheck.abi0")
-	if err != nil {
-		// Binary must be stripped, hope default is correct and warn.
-		log.Warnf("Failed to find stackcheck symbol, Go labels might not work: %v", err)
-		return -8, nil
+	// Case 1 - direct offset
+	//   mov %fs:0xfffffffffffffff8,%rax
+	// Case 2 - indirect offset
+	//   mov 0x17d0c5a1(%rip),%rcx
+	//   mov %fs:(%rcx),%rax
+	pc, ok := pclntab.PCForSymbol("runtime.stackcheck")
+	if !ok {
+		return 0, errMissingSymbol
 	}
-	b, err := f.VirtualMemory(int64(sym.Address), 10, 10)
+	// Read enough bytes for two instructions
+	b, err := f.VirtualMemory(int64(pc), 16, 16)
 	if err != nil {
 		return 0, err
 	}
+	return extractOffsetFromBytes(f, pc, b)
+}
 
-	i, err := x86asm.Decode(b, 64)
+func extractOffsetFromBytes(f virtualMemoryReader, pc uintptr, b []byte) (int32, error) {
```

Comment:
See #730. I think it handles more this alternative instruction sequence in a more complete way including arm64. Also the amd64 side looks cleaner as it uses the new disassembly machinery. It also does not do additional `VirtualMemory` reads as it tracks the registers. So the `virtualMemoryReader` interface for this function is not needed.